File: uct.ml
type reward_t = float
type 's problem = 's -> ('s * reward_t) list

type 's tree = {
  state : 's
; visits : int
; rewards : reward_t
; children : ('s tree * reward_t) list
; actions : ('s * reward_t) list
}

let empty_tree problem s = {
  state = s
; visits = 0
; rewards = 0.
; children = []
; actions = List.sort (fun (_, x) (_, y) -> compare y x) (problem s)
}

let avg_reward tree = tree.rewards /. float_of_int tree.visits

let child_score parent child =
  let c_p = sqrt 2. and fi = float_of_int in
  avg_reward child +. c_p *. sqrt (log (fi parent.visits) /. fi child.visits)

let child_cmp parent (c1, _) (c2, _) =
  compare (child_score parent c2) (child_score parent c1)

let rec search p tree = match tree.actions with
    (act, rew) :: acts -> update p (empty_tree p act, rew) {tree with actions = acts}
  | [] ->
      (match List.sort (child_cmp tree) tree.children with
        [] -> {tree with visits = tree.visits + 1}
      | best :: rest -> update p best {tree with children = rest})
and update p (child, rew) parent =
  let gamma = 0.7
  and child' = search p child in
  { parent with
    rewards = parent.rewards +. rew +. gamma *. (child'.rewards -. child.rewards)
  ; visits = parent.visits + 1
  ; children = (child', rew) :: parent.children
  }

let by_reward (c1, _) (c2, _) = compare (avg_reward c2) (avg_reward c1)

let by_visits (c1, _) (c2, _) = compare c2.visits c1.visits

let rec best_state cmp tree = match List.sort cmp tree.children with
    (best, _) :: _ -> best_state cmp best
  | [] -> tree.state

File: UCT.hs
module UCT where

import Data.List (sortOn)
import Data.Ord (Down(..))
import Debug.Trace

type Reward = Double
type Problem s = s -> [(s, Reward)]

data Tree s = Tree {
  state :: s
, visits :: Int
, rewards :: Reward
, children :: [(Tree s, Reward)]
, actions :: [(s, Reward)]
} deriving Show

emptyTree :: Problem s -> s -> Tree s
emptyTree problem s = Tree {
  state = s
, visits = 0
, rewards = 0
, children = []
, actions = sortOn (Down . snd) (problem s)
}

avgReward :: Tree s -> Reward
avgReward tree = rewards tree / fromIntegral (visits tree)

childScore :: Tree s -> Tree s -> Double
childScore parent child = avgReward child + c_p * sqrt (log t / s)
  where c_p = 0
        t = fromIntegral (visits parent)
        s = fromIntegral (visits child)

search :: Problem s -> Tree s -> Tree s
search p tree = case actions tree of
  (act, rew) : acts -> update p (emptyTree p act, rew) (tree {actions = acts})
  [] -> case sortOn (Down . childScore tree . fst) (children tree) of
    [] -> tree {visits = visits tree + 1}
    best : rest -> update p best (tree {children = rest})

update :: Problem s -> (Tree s, Reward) -> Tree s -> Tree s
update p (child, rew) parent = parent
  { rewards = rewards parent + rew + gamma * (rewards child' - rewards child)
  , visits = visits parent + 1
  , children = (child', rew) : children parent
  }
  where gamma = 1
        child' = search p child

bestState :: Ord a => (Tree s -> a) -> Tree s -> s
bestState f tree = case sortOn (f . fst) (children tree) of
  (best, _) : _ -> bestState f best
  [] -> state tree

byReward :: Tree s -> Down Reward
byReward = Down . avgReward

byVisits :: Tree s -> Down Int
byVisits = Down . visits

normRewards l = map (\ (s, r) -> (s, (r - min) / (max - min))) l
  where max = maximum (map snd l)
        min = minimum (map snd l)

